File: app/src/main/java/io/treehouses/remote/Fragments/ServiceCardFragment.java
package io.treehouses.remote.Fragments;

import android.content.Context;
import android.graphics.drawable.PictureDrawable;
import android.os.Bundle;
import android.text.SpannableString;
import android.text.method.LinkMovementMethod;
import android.text.util.Linkify;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;

import com.caverock.androidsvg.SVG;
import com.caverock.androidsvg.SVGParseException;

import io.treehouses.remote.callback.ServiceAction;
import io.treehouses.remote.databinding.ServiceCardBinding;
import io.treehouses.remote.pojo.ServiceInfo;

public class ServiceCardFragment extends Fragment implements View.OnClickListener {
    private ServiceInfo serviceData;
    private ServiceAction actionListener;

    private ServiceCardBinding binding;

    public ServiceCardFragment(ServiceInfo serviceData) {
        this.serviceData = serviceData;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = ServiceCardBinding.inflate(inflater, container, false);

        binding.serviceInfo.setMovementMethod(LinkMovementMethod.getInstance());
        binding.serviceInfo.setFocusable(true);


        if (!serviceData.isHeader()) {
            setServiceInfo(serviceData.info);
            showIcon(serviceData.icon);
            updateButtons(serviceData.serviceStatus);
            setAutorun(serviceData.autorun);

            binding.installButton.setOnClickListener(this);
            binding.startButton.setOnClickListener(this);
            binding.openLink.setOnClickListener(this);
            binding.autorunChecked.setOnCheckedChangeListener((buttonView, isChecked) -> actionListener.onClickAutorun(serviceData, isChecked));
        }
        return binding.getRoot();
    }

    private void setAutorun(String autorun) {
        binding.autorunChecked.setChecked(autorun.contains("true"));
    }

    private void setButtons(boolean started, boolean installed, boolean three) {
        if (started) {
            binding.startButton.setText("Stop");
            binding.openLink.setVisibility(View.VISIBLE);
        }
        else {
            binding.startButton.setText("Start");
            binding.openLink.setVisibility(View.GONE);
        }
        if (installed) {
            binding.installButton.setText("Uninstall");
            binding.startButton.setVisibility(View.VISIBLE);
            binding.autorunChecked.setVisibility(View.VISIBLE);
        } else {
            binding.installButton.setText("Install");
            binding.startButton.setVisibility(View.GONE);
            binding.autorunChecked.setVisibility(View.GONE);
        }
        //restart.setEnabled(three);
    }

    private void updateButtons(int statusCode) {
        switch (statusCode) {
            case ServiceInfo.SERVICE_AVAILABLE:
                setButtons(false, false, false);
                break;
            case ServiceInfo.SERVICE_INSTALLED:
                setButtons(false, true, false);
                break;
            case ServiceInfo.SERVICE_RUNNING:
                setButtons(true, true, true);
                break;
        }
    }

    private void showIcon(String s) {
        try {
            Log.d(serviceData.name, "showIcon:" + serviceData.icon);
            SVG svg = SVG.getFromString(s);
            PictureDrawable pd = new PictureDrawable(svg.renderToPicture());
            binding.serviceLogo.setImageDrawable(pd);
        } catch (SVGParseException e) {
            e.printStackTrace();
        }
    }

    private void setServiceInfo(String s) {
        SpannableString spannableString = new SpannableString(s);
        Linkify.addLinks(spannableString, Linkify.ALL);
        binding.serviceInfo.setText(s);
        binding.serviceInfo.setMovementMethod(LinkMovementMethod.getInstance());
    }

//    private void onClickRestart(ServiceInfo selected) {
//        if (selected.serviceStatus != ServiceInfo.SERVICE_AVAILABLE) performService("Restarting", "treehouses services " + selected.name + " restart\n", selected.name);
//
//    }
//    private void setOnClick(View v, int id, String command, AlertDialog alertDialog) {
//        v.findViewById(id).setOnClickListener(new View.OnClickListener() {
//            @Override
//            public void onClick(View v) {
//                writeToRPI(command);
//                alertDialog.dismiss();
////                progressBar.setVisibility(View.VISIBLE);
//            }
//        });
//    }


    @Override
    public void onClick(View v) {
        if (binding.installButton.equals(v)) {
            actionListener.onClickInstall(serviceData);
        } else if (binding.startButton.equals(v)) {
            actionListener.onClickStart(serviceData);
        } else if (binding.openLink.equals(v)) {
            actionListener.onClickLink(serviceData);
        }
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        actionListener = (ServiceAction) getParentFragment();
    }

}

File: app/src/main/kotlin/io/treehouses/remote/fragments/ServiceCardFragment.kt
package io.treehouses.remote.fragments

import android.content.Context
import android.graphics.drawable.PictureDrawable
import android.os.Bundle
import android.text.SpannableString
import android.text.method.LinkMovementMethod
import android.text.util.Linkify
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import com.caverock.androidsvg.SVG
import io.treehouses.remote.callback.ServiceActionListener
import io.treehouses.remote.databinding.ServiceCardBinding
import io.treehouses.remote.pojo.ServiceInfo

class ServiceCardFragment : Fragment(), View.OnClickListener {
    private var actionListener: ServiceActionListener? = null
    private var binding: ServiceCardBinding? = null
    private lateinit var serviceData: ServiceInfo

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        arguments?.let {
            if (it.containsKey("serviceData"))
                serviceData = it.getSerializable("serviceData") as ServiceInfo
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        binding = ServiceCardBinding.inflate(inflater, container, false)
        binding!!.serviceInfo.movementMethod = LinkMovementMethod.getInstance()
        binding!!.serviceInfo.isFocusable = true
        if (!serviceData.isHeader) {
            setServiceInfo(serviceData.info)
            setServiceSize(serviceData.size)
            showIcon(serviceData.icon)
            updateButtons(serviceData.serviceStatus)
            setAutorun(serviceData.autorun)
            binding!!.installButton.setOnClickListener(this)
            binding!!.startButton.setOnClickListener(this)
            binding!!.openLink.setOnClickListener(this)
            binding!!.editEnvButton.setOnClickListener(this)
            binding!!.autorunChecked.setOnClickListener {
                actionListener!!.onClickAutorun(serviceData, binding?.autorunChecked?.isChecked ?: false)
            }
        }
        return binding!!.root
    }

    private fun setAutorun(autorun: String?) {
        binding!!.autorunChecked.isChecked = autorun!!.contains("true")
    }

    private fun setButtons(started: Boolean, installed: Boolean) {
        var string1 = "Start"
        var visibility1 = View.GONE

        if (started) {
            string1 = "Stop"
            visibility1 = View.VISIBLE
        }

        var string2 = "Install"
        var visibility2 = View.GONE
        var visibility3 = View.GONE

        binding!!.startButton.text = string1
        binding!!.openLink.visibility = visibility1

        if (installed) {
            string2 = "Uninstall"
            visibility2 = View.VISIBLE
            visibility3 = View.VISIBLE
        }
        var visibility4 = View.GONE

        if (installed && !started && serviceData.usesEnv == "true")
            visibility4 = View.VISIBLE

        binding!!.installButton.text = string2
        binding!!.startButton.visibility = visibility2
        binding!!.autorunChecked.visibility = visibility3
        binding!!.editEnvButton.visibility = visibility4

    }

    private fun updateButtons(statusCode: Int) {
        when (statusCode) {
            ServiceInfo.SERVICE_AVAILABLE -> setButtons(started = false, installed = false)
            ServiceInfo.SERVICE_INSTALLED -> setButtons(started = false, installed = true)
            ServiceInfo.SERVICE_RUNNING -> setButtons(started = true, installed = true)
        }
    }

    private fun showIcon(s: String?) {
        try {
            val svg = SVG.getFromString(s)
            val pd = PictureDrawable(svg.renderToPicture())
            binding!!.serviceLogo.setImageDrawable(pd)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun setServiceInfo(s: String?) {
        val spannableString = SpannableString(s)
        Linkify.addLinks(spannableString, Linkify.ALL)
        binding!!.serviceInfo.text = spannableString
        binding!!.serviceInfo.movementMethod = LinkMovementMethod.getInstance()
    }

    private fun setServiceSize(sz: Int) {
        if(sz != -1)
            binding!!.serviceSize.text = "$sz MB"
    }

    override fun onClick(v: View) {
        when {
            binding!!.installButton == v -> {
                actionListener!!.onClickInstall(serviceData)
            }
            binding!!.startButton == v -> {
                actionListener!!.onClickStart(serviceData)
            }
            binding!!.openLink == v -> {
                actionListener!!.onClickLink(serviceData)
            }
            binding!!.editEnvButton == v -> {
                actionListener!!.onClickEditEnvVar(serviceData)
            }
        }
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        actionListener = parentFragment as ServiceActionListener?
    }

}